cmd/kops/create_cluster.go

```diff
@@ -441,8 +458,15 @@ func RunCreateCluster(f *util.Factory, out io.Writer, c *CreateClusterOptions) e
 		return fmt.Errorf("unknown authorization mode %q", c.Authorization)
 	}
 
+	var cloudProvider api.CloudProviderID
 	if c.Cloud != "" {
 		cluster.Spec.CloudProvider = c.Cloud
+		cloudProvider = api.CloudProviderID(c.Cloud)
+
+		if cloudProvider == api.CloudProviderSpotinst {
```

Comment:
What about combining these as it is an edge case? 
```go
if c.Cloud != "" && api.CloudProviderID(c.Cloud) == api.CloudProviderSpotinst {
```

Would also reduce the change footprint down the line.

Reply:
Not sure it's possible. On the one hand, we have to set the `cluster.Spec.CloudPovider` to be `spotinst`, but on the other hand, we have to use a [separate variable](https://github.com/spotinst/kubernetes-kops/blob/64176c7274f52af7408b28bf3be72cc399e3a7b1/cmd/kops/create_cluster.go#L461) (`cloudProvider`) so we can change it to the real cloud provider and then use it later, for example, [to set the VPC ID](https://github.com/spotinst/kubernetes-kops/blob/64176c7274f52af7408b28bf3be72cc399e3a7b1/cmd/kops/create_cluster.go#L478) in case of the underlying cloud provider is AWS.

Reply:
My proposal would be - for all steps - keep the cloud provider as aws and extend the providers with a flag that it’s to be used by spotinst. That way you “only” need to fetch the processed specs where we send it to the provider and you implement a switch to send it to spotinst provider instead. 

That way we can keep everything clean and just send all things to spotinst - which can modify the small bits required and send it to their api then.

Reply:
I understand, but this change will require us to change the AWS implementation, which we tried to avoid. Do you want us to do this? I can try to work on it.

Reply:
My approach would be just the hook before the AWS implementation starts. Meaning INSTEAD of going INTO the AWS implementation we trigger a new spotinst implementation which then does all it requires to implement all the clouds. 

Just that we keep all the input for the providers the same - and the spotinst implementation - at this point - transforms it to a compatible state and sends it to the spotinst endpoints.

Reply:
Thats what I mean with a special spotinst part in nodeup. If it is setup to use nodeup instead of AWS we replace the nodeup parts from AWS with a spotinst alternative that then talks to different API's. 

I'm not talking about keeping AWS there - I'm just talking about keeping the way we configure kops the same and just map that to spotinst patterns and send it to different API's as a REPLACEMENT in nodeup phase for aws, gcloud and such. 

That keeps our configuration model, wording and everything consistent - just uses other resources in the end. But - after all - the concepts remain the same. Just the actual service is named differently.

Reply:
I thought you were talking about setting up AWS as a cloud provider and using Spotinst-specific flags to set up resources differently but maybe I'm missing something since I'm not understanding what you're offering. Really sorry about that. It would be great if you could refer me to the relevant part you are talking about.

Reply:
https://github.com/kubernetes/kops/blob/master/upup/pkg/fi/cloudup/apply_cluster.go#L362

I'm talking about this place. If we "switch" into spotinst mode here we can keep everything in kops the same, just apply it to the spotinst api in replacement to the AWS api.

Reply:
@liranp as we talked about before having another compete cloud provider is quite challenging.  I understand your clients will need to migrate, but it should be a simple change for them to do so.  We can setup a time to talk if you like or I can try to make the next developer office hours.  We really need to utilize the existing aws cloud provider and either have another name or a flag that the cloud provider can check.

Reply:
@chrislovecnm We are currently using the AWS implementation, but planning to configure the cloud provider to be `external` and run our own cloud-controller-manager that will use AWS (or any other cloud provider) implementation with changes, such as overriding the LoadBalancer part that will use our product instead of AWS/ELB. Do you think we can do this if we do not become an independent cloud provider?

Reply:
@liranp that would be possible still. You just switch out the image for the cloud controller manager in nodeup as well.